File: client/src/pages/Game/shared/itemTaxonomy.ts
/**
 * 物品分类/子类公用字典
 *
 * 作用：
 * - 统一背包、自动分解等模块对“主分类/子类”枚举与中文文案的单一来源，避免各模块独自维护导致口径漂移。
 * - 提供稳定可复用的 options 与按分类分组的子类白名单，减少重复映射逻辑。
 *
 * 输入/输出：
 * - 输入：无（纯静态配置模块）。
 * - 输出：
 *   1) `BagCategory`、`BagPrimaryCategory` 类型；
 *   2) 主分类文案、主分类 options；
 *   3) 子类 options、子类按主分类白名单。
 *
 * 数据流/状态流：
 * - 业务模块仅消费本模块导出的常量与类型，不在模块外改写；
 * - 自动分解与背包筛选都复用同一字典，实现“改一处、全链路同步”。
 *
 * 关键边界条件与坑点：
 * 1) `all` 仅用于界面筛选态，不应进入自动分解规则持久化数据。
 * 2) `technique_book` 与 `technique` 必须保持在 `skill` 主分类白名单中，避免功法相关筛选失效。
 */

export interface LabeledOption {
  label: string;
  value: string;
}

export type BagCategory =
  | 'all'
  | 'consumable'
  | 'material'
  | 'gem'
  | 'equipment'
  | 'skill'
  | 'quest';

export type BagPrimaryCategory = Exclude<BagCategory, 'all'>;

export const BAG_CATEGORY_LABELS: Record<BagCategory, string> = {
  all: '全部',
  consumable: '丹药',
  material: '材料',
  gem: '宝石',
  equipment: '装备',
  skill: '功法',
  quest: '任务',
};

const BAG_CATEGORY_OPTION_ORDER: BagPrimaryCategory[] = [
  'equipment',
  'consumable',
  'material',
  'gem',
  'skill',
  'quest',
];

export const BAG_CATEGORY_OPTIONS: LabeledOption[] = BAG_CATEGORY_OPTION_ORDER.map((value) => ({
  value,
  label: BAG_CATEGORY_LABELS[value],
}));

export const BAG_SUB_CATEGORY_OPTIONS: LabeledOption[] = [
  { label: '剑', value: 'sword' },
  { label: '刀', value: 'blade' },
  { label: '法杖', value: 'staff' },
  { label: '盾牌', value: 'shield' },
  { label: '头盔', value: 'helmet' },
  { label: '帽子', value: 'hat' },
  { label: '法袍', value: 'robe' },
  { label: '手套', value: 'gloves' },
  { label: '臂甲', value: 'gauntlets' },
  { label: '下装', value: 'pants' },
  { label: '护腿', value: 'legguards' },
  { label: '戒指', value: 'ring' },
  { label: '项链', value: 'necklace' },
  { label: '护符', value: 'talisman' },
  { label: '宝镜', value: 'mirror' },
  { label: '配饰', value: 'accessory' },
  { label: '护甲', value: 'armor' },
  { label: '战令道具', value: 'battle_pass' },
  { label: '骨材', value: 'bone' },
  { label: '宝箱', value: 'box' },
  { label: '突破道具', value: 'breakthrough' },
  { label: '采集物', value: 'collect' },
  { label: '蛋类', value: 'egg' },
  { label: '强化道具', value: 'enhance' },
  { label: '精华', value: 'essence' },
  { label: '锻造材料', value: 'forge' },
  { label: '功能道具', value: 'function' },
  { label: '宝石', value: 'gem' },
  { label: '攻击宝石', value: 'gem_attack' },
  { label: '防御宝石', value: 'gem_defense' },
  { label: '生存宝石', value: 'gem_survival' },
  { label: '通用宝石', value: 'gem_all' },
  { label: '灵草', value: 'herb' },
  { label: '钥匙', value: 'key' },
  { label: '皮革', value: 'leather' },
  { label: '月卡道具', value: 'month_card' },
  { label: '杂项道具', value: 'object' },
  { label: '矿石', value: 'ore' },
  { label: '丹药', value: 'pill' },
  { label: '遗物', value: 'relic' },
  { label: '卷轴', value: 'scroll' },
  { label: '功法', value: 'technique' },
  { label: '功法书', value: 'technique_book' },
  { label: '代币', value: 'token' },
  { label: '木材', value: 'wood' },
];

export const BAG_SUB_CATEGORY_VALUES_BY_CATEGORY: Record<BagCategory, string[]> = {
  all: BAG_SUB_CATEGORY_OPTIONS.map((option) => option.value),
  consumable: ['pill', 'box', 'function', 'enhance', 'scroll', 'month_card', 'battle_pass', 'token'],
  material: [
    'herb',
    'ore',
    'wood',
    'leather',
    'essence',
    'bone',
    'relic',
    'forge',
    'breakthrough',
    'egg',
    'accessory',
    'armor',
    'object',
  ],
  gem: ['gem', 'gem_attack', 'gem_defense', 'gem_survival', 'gem_all'],
  equipment: [
    'sword',
    'blade',
    'staff',
    'shield',
    'helmet',
    'hat',
    'robe',
    'gloves',
    'gauntlets',
    'pants',
    'legguards',
    'ring',
    'necklace',
    'talisman',
    'mirror',
    'accessory',
    'armor',
    'token',
  ],
  skill: ['technique', 'technique_book'],
  quest: ['key', 'collect'],
};

File: server-rs/src/repo/item_taxonomy.rs
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};

use crate::shared::error::AppError;

#[derive(Debug, Clone, Serialize)]
pub struct ItemTaxonomyOptionDto {
    pub value: String,
    pub label: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GameItemTaxonomyDto {
    pub categories: GameItemTaxonomyCategorySection,
    pub sub_categories: GameItemTaxonomySubCategorySection,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GameItemTaxonomyCategorySection {
    pub all: ItemTaxonomyOptionDto,
    pub options: Vec<ItemTaxonomyOptionDto>,
    pub labels: BTreeMap<String, String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GameItemTaxonomySubCategorySection {
    pub options: Vec<ItemTaxonomyOptionDto>,
    pub labels: BTreeMap<String, String>,
    pub by_category: BTreeMap<String, Vec<String>>,
}

#[derive(Debug, Deserialize)]
struct ItemDefFile {
    items: Vec<ItemDefSeed>,
}

#[derive(Debug, Deserialize)]
struct ItemDefSeed {
    id: Option<String>,
    category: Option<String>,
    sub_category: Option<String>,
    enabled: Option<bool>,
}

const CATEGORY_PREFERRED_ORDER: &[&str] = &[
    "consumable",
    "material",
    "gem",
    "equipment",
    "quest",
    "other",
];

pub fn build_game_item_taxonomy() -> Result<GameItemTaxonomyDto, AppError> {
    let seeds = load_merged_item_seeds()?;
    let mut category_set = BTreeSet::new();
    let mut sub_category_labels = BTreeMap::new();
    let mut by_category: BTreeMap<String, Vec<String>> =
        BTreeMap::from([("all".to_string(), Vec::new())]);

    for item in seeds
        .into_iter()
        .filter(|entry| entry.enabled != Some(false))
    {
        let category = normalize_token(item.category.as_deref());
        if !category.is_empty() {
            category_set.insert(category.clone());
            by_category.entry(category.clone()).or_default();
        }

        let sub_category = normalize_token(item.sub_category.as_deref());
        if sub_category.is_empty() {
            continue;
        }

        sub_category_labels
            .entry(sub_category.clone())
            .or_insert_with(|| sub_category_label(sub_category.as_str()));
        push_unique(
            by_category.entry("all".to_string()).or_default(),
            sub_category.clone(),
        );
        if !category.is_empty() {
            push_unique(by_category.entry(category).or_default(), sub_category);
        }
    }

    let category_values = sort_category_values(category_set.into_iter().collect());
    let mut category_labels = BTreeMap::from([("all".to_string(), "全部".to_string())]);
    for category in &category_values {
        category_labels.insert(category.clone(), category_label(category));
        by_category.entry(category.clone()).or_default();
    }

    let category_options = category_values
        .iter()
        .map(|value| ItemTaxonomyOptionDto {
            value: value.clone(),
            label: category_labels
                .get(value)
                .cloned()
                .unwrap_or_else(|| value.clone()),
        })
        .collect();

    let mut sub_category_options: Vec<ItemTaxonomyOptionDto> = sub_category_labels
        .iter()
        .map(|(value, label)| ItemTaxonomyOptionDto {
            value: value.clone(),
            label: label.clone(),
        })
        .collect();
    sub_category_options.sort_by(|left, right| {
        left.label
            .cmp(&right.label)
            .then_with(|| left.value.cmp(&right.value))
    });

    Ok(GameItemTaxonomyDto {
        categories: GameItemTaxonomyCategorySection {
            all: ItemTaxonomyOptionDto {
                value: "all".to_string(),
                label: "全部".to_string(),
            },
            options: category_options,
            labels: category_labels,
        },
        sub_categories: GameItemTaxonomySubCategorySection {
            options: sub_category_options,
            labels: sub_category_labels,
            by_category: by_category,
        },
    })
}

fn read_item_seed_file(filename: &str) -> Result<ItemDefFile, AppError> {
    let path = item_seed_path(filename);
    let content = fs::read_to_string(&path)
        .map_err(|error| AppError::config(format!("failed to read {filename}: {error}")))?;
    serde_json::from_str(&content)
        .map_err(|error| AppError::config(format!("failed to parse {filename}: {error}")))
}

fn item_seed_path(filename: &str) -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(format!("../server/src/data/seeds/{filename}"))
}

fn load_merged_item_seeds() -> Result<Vec<ItemDefSeed>, AppError> {
    let mut by_id: BTreeMap<String, ItemDefSeed> = BTreeMap::new();
    for filename in ["item_def.json", "gem_def.json", "equipment_def.json"] {
        let file = read_item_seed_file(filename)?;
        for item in file.items {
            let id = normalize_token(item.id.as_deref());
            if id.is_empty() {
                continue;
            }
            by_id.insert(id, item);
        }
    }
    Ok(by_id.into_values().collect())
}

fn normalize_token(raw: Option<&str>) -> String {
    raw.unwrap_or_default().trim().to_lowercase()
}

fn push_unique(list: &mut Vec<String>, value: String) {
    if !list.contains(&value) {
        list.push(value);
    }
}

fn sort_category_values(mut values: Vec<String>) -> Vec<String> {
    values.sort_by(|left, right| {
        let left_rank = CATEGORY_PREFERRED_ORDER
            .iter()
            .position(|value| *value == left);
        let right_rank = CATEGORY_PREFERRED_ORDER
            .iter()
            .position(|value| *value == right);
        match (left_rank, right_rank) {
            (Some(l), Some(r)) => l.cmp(&r),
            (Some(_), None) => std::cmp::Ordering::Less,
            (None, Some(_)) => std::cmp::Ordering::Greater,
            (None, None) => left.cmp(right),
        }
    });
    values
}

fn category_label(value: &str) -> String {
    match value {
        "consumable" => "消耗品",
        "material" => "材料",
        "gem" => "宝石",
        "equipment" => "装备",
        "quest" => "任务",
        "other" => "其他",
        _ => value,
    }
    .to_string()
}

fn sub_category_label(value: &str) -> String {
    match value {
        "sword" => "剑",
        "blade" => "刀",
        "staff" => "法杖",
        "shield" => "盾牌",
        "helmet" => "头盔",
        "hat" => "帽子",
        "robe" => "法袍",
        "gloves" => "手套",
        "gauntlets" => "臂甲",
        "pants" => "下装",
        "legguards" => "护腿",
        "ring" => "戒指",
        "necklace" => "项链",
        "talisman" => "法宝（护符）",
        "mirror" => "宝镜",
        "accessory" => "饰品",
        "armor" => "护甲",
        "battle_pass" => "战令道具",
        "bone" => "骨材",
        "box" => "宝箱",
        "breakthrough" => "突破道具",
        "collect" => "采集物",
        "egg" => "蛋类",
        "enhance" => "强化道具",
        "essence" => "精华",
        "forge" => "锻造材料",
        "function" => "功能道具",
        "gem" => "宝石",
        "gem_attack" => "攻击宝石",
        "gem_defense" => "防御宝石",
        "gem_survival" => "生存宝石",
        "gem_all" => "通用宝石",
        "herb" => "灵草",
        "key" => "钥匙",
        "leather" => "皮革",
        "month_card" => "月卡道具",
        "object" => "杂项道具",
        "ore" => "矿石",
        "pill" => "丹药",
        "relic" => "遗物",
        "scroll" => "卷轴",
        "technique" => "功法材料",
        "technique_book" => "功法书",
        "token" => "法宝",
        "wood" => "木材",
        _ => value,
    }
    .to_string()
}

#[cfg(test)]
mod tests {
    #[test]
    fn taxonomy_reads_seed_file() {
        let taxonomy = super::build_game_item_taxonomy().expect("taxonomy should build");
        assert!(!taxonomy.categories.options.is_empty());
        assert!(taxonomy.sub_categories.labels.contains_key("pill"));
        assert!(taxonomy.categories.labels.contains_key("equipment"));
        assert!(taxonomy.sub_categories.by_category.contains_key("gem"));
    }
}
